src/daft-io/src/object_io.rs

```diff
@@ -112,10 +117,203 @@ pub(crate) trait ObjectSource: Sync + Send {
+) -> super::Result<BoxStream<'static, super::Result<FileMetadata>>> {
+    let glob_fragments = to_glob_fragments(glob)?;
+    let full_glob_matcher = GlobBuilder::new(glob)
+        .literal_separator(true)
+        .backslash_escape(true)
+        .build()
+        .map_err(|err| super::Error::InvalidArgument {
+            msg: format!("Cannot parse provided glob {glob}: {err}"),
+        })?
+        .compile_matcher();
+
+    // Channel to send results back to caller. Note that all results must have FileType::File.
+    let (to_rtn_tx, mut to_rtn_rx) = tokio::sync::mpsc::channel(16 * 1024);
+
+    /// Dispatches a task to visit the specified `path` (a concrete path on the filesystem to either a File or Directory).
+    /// Based on the current glob_fragment being processed (accessible via `glob_fragments[i]`) this task will:
+    ///   1. Perform work to retrieve Files/Directories at (`path` + `glob_fragments[i]`)
+    ///   2. Return results to the provided `result_tx` channel based on the provided glob, if appropriate
+    ///   3. Dispatch additional tasks via `.visit()` to continue visiting them, if appropriate
+    fn visit(
+        result_tx: Sender<super::Result<FileMetadata>>,
+        source: Arc<dyn ObjectSource>,
+        path: &str,
+        glob_fragments: (Arc<Vec<GlobFragment>>, usize),
+        full_glob_matcher: Arc<GlobMatcher>,
+    ) {
+        let path = path.to_string();
+        tokio::spawn(async move {
+            log::debug!(target: "glob", "Visiting '{path}' with glob_fragments: {glob_fragments:?}");
+            let (glob_fragments, i) = glob_fragments;
+            let current_fragment = &glob_fragments[i];
+
+            // BASE CASE: current_fragment is a **
+            // We perform a recursive ls and filter on the results for only FileType::File results that match the full glob
+            if current_fragment.escaped_str() == "**" {
+                let next_level_file_metadata =
+                    source.iter_dir(path.as_str(), Some("/"), None).await;
+
+                match next_level_file_metadata {
+                    Ok(mut next_level_file_metadata) => {
+                        while let Some(fm) = next_level_file_metadata.next().await {
+                            match fm {
+                                Ok(fm) => {
+                                    // Recursively visit each sub-directory, do not increment `i` so as to keep visiting the "**" fragmemt
+                                    if matches!(fm.filetype, FileType::Directory) {
+                                        visit(
+                                            result_tx.clone(),
+                                            source.clone(),
+                                            &fm.filepath,
+                                            (glob_fragments.clone(), i),
+                                            full_glob_matcher.clone(),
+                                        );
+                                    }
```

Comment:
can these both be true, if not, you can do else if or a match on the `FileType`

Reply:
Could you elaborate on this comment?

Reply:
Oh! I understand. Fixing to use a match.